syntax = "proto3";

import "pipelines_and_context.proto";
import "util.proto";
import "backend_service.proto";
import "google/protobuf/empty.proto";

package ficus;

service GrpcKafkaService {
  rpc SubscribeForKafkaTopic(GrpcSubscribeToKafkaRequest) returns (GrpcKafkaResult);
  rpc UnsubscribeFromKafkaTopic(GrpcUnsubscribeFromKafkaRequest) returns (GrpcKafkaResult);
  
  rpc AddPipelineToSubscription(GrpcAddPipelineRequest) returns (GrpcKafkaResult);
  rpc AddPipelineToSubscriptionStream(GrpcAddPipelineStreamRequest) returns (stream GrpcPipelinePartExecutionResult);

  rpc RemovePipelineSubscription(GrpcRemovePipelineRequest) returns (GrpcKafkaResult);
  rpc RemoveAllPipelineSubscriptions(GrpcRemoveAllPipelinesRequest) returns (GrpcKafkaResult);

  rpc ExecutePipelineAndProduceToKafka(GrpcExecutePipelineAndProduceKafkaRequest) returns (stream GrpcPipelinePartExecutionResult);
}

message GrpcExecutePipelineAndProduceKafkaRequest {
  GrpcProxyPipelineExecutionRequest pipelineRequest = 1;
  GrpcKafkaConnectionMetadata producerMetadata = 2;
  GrpcProcessInfo caseInfo = 3;
}

message GrpcProcessInfo {
  string processName = 1;
  string caseName = 2;
}

message GrpcSubscribeToKafkaRequest {
  GrpcKafkaConnectionMetadata connectionMetadata = 1;
  string subscriptionName = 2;
}

message GrpcKafkaPipelineExecutionRequest {
  GrpcGuid subscriptionId = 1;
  GrpcPipelineExecutionRequest pipelineRequest = 2;
  string pipelineName = 3;
}

message GrpcAddPipelineRequest {
  GrpcKafkaPipelineExecutionRequest pipelineRequest = 1;
  GrpcKafkaConnectionMetadata producerKafkaMetadata = 2;
}

message GrpcAddPipelineStreamRequest {
  GrpcKafkaPipelineExecutionRequest pipelineRequest = 1;
}

message GrpcRemovePipelineRequest {
  GrpcGuid subscriptionId = 1;
  GrpcGuid pipelineId = 2;
}

message GrpcRemoveAllPipelinesRequest {
  GrpcGuid subscriptionId = 1;
}

message GrpcKafkaConnectionMetadata {
  string topicName = 1;
  repeated GrpcKafkaConsumerMetadata metadata = 2;
}

message GrpcKafkaConsumerMetadata {
  string key = 1;
  string value = 2;
}

message GrpcKafkaResult {
  oneof result {
    GrpcKafkaSuccessResult success = 1;
    GrpcKafkaFailedResult failure = 2;
  }
}

message GrpcKafkaSuccessResult {
  GrpcGuid id = 1;
}

message GrpcKafkaFailedResult {
  string errorMessage = 1;
}

message GrpcUnsubscribeFromKafkaRequest {
  GrpcGuid subscriptionId = 1;
}